### PR TITLE
tree wide use ensure_put

### PR DESCRIPTION
**Pull request by ssahani**

### CODE REVIEW

File: src/network/netdev/macsec.c [hunk 1]
```diff
@@ -157,11 +153,7 @@ static int macsec_receive_channel_new_static(MACsec *s, const char *filename, un
 
         c->section = TAKE_PTR(n);
 
-        r = ordered_hashmap_ensure_allocated(&s->receive_channels_by_section, &network_config_hash_ops);
-        if (r < 0)
-                return r;
-
-        r = ordered_hashmap_put(s->receive_channels_by_section, c->section, c);
+        r = ordered_hashmap_ensure_put(&s->receive_channels_by_section, &network_config_hash_ops, c->section, c);
```
Comment (bluca) [inline src/network/netdev/macsec.c:156]: any particular reason these 2 don't get a specific log_oom but the rest does?

Comment (ssahani) [reply src/network/netdev/macsec.c:156]: We can add that. It was not there before.

**Review by bluca — APPROVED** [review]
lgtm

**Comment by ssahani** [comment]
Thanks for the review.

File: src/network/netdev/macsec.c [hunk 2]
```diff
@@ -216,11 +210,9 @@ static int macsec_transmit_association_new_static(MACsec *s, const char *filenam
...
         security_association_init(&a->sa);
 
-        r = ordered_hashmap_ensure_allocated(&s->transmit_associations_by_section, &network_config_hash_ops);
-        if (r < 0)
-                return r;
-
-        r = ordered_hashmap_put(s->transmit_associations_by_section, a->section, a);
+        r = ordered_hashmap_ensure_put(&s->transmit_associations_by_section, &network_config_hash_ops, a->section, a);
+        if (r == -ENOMEM)
+                return log_oom();
```
Comment (poettering) [inline src/network/netdev/macsec.c]: this function so far was one that doesn't log about errors, but let's the caller do this, afaics.  hence, it shouldn't log on ENOMEM now here too. 

i.e. our functions should either be of the "logging kind" or of the "non-logging kind". i.e. log about all errors they encounter, or about none (above LOG_DEBUG log level that is) — in the latter case the caller can do that instead.

So far, macsec_transmit_association_new_static() afaics is of the non-logging kind, and hence shouldn't log here either.

macsec_receive_channel_verify() otoh seems to be of the logging kind, hence everything is in order there, with the new log_oom() being inserted.

(yes, we don't get this right in many cases, and that should be fixed sooner or later. Let's fix this whenever we touch code though, as we do here...)

**Review by poettering — CHANGES_REQUESTED** [review]
Cleaning this up definitely makes sense, but we should try to get the logging-vs-non-logging function stuff right when doing so. i.e. as mentioned below: please do the log_oom() thing only in functions that actually are of the "logging kind". i.e. please identify first what kind of function you are looking at, then log on all errors if it's of the "logging king", and on none if it's of the "non-logging kind" (non-logging means: no logging above LOG_DEBUG — debug logging is always OK).

File: src/network/netdev/macsec.c (hunk 2 above)
Comment (ssahani) [reply src/network/netdev/macsec.c]: OK thanks reverted.

Comment (yuwata) [reply src/network/netdev/macsec.c]: Not addressed yet...

File: src/network/networkd-address.c [hunk 3]
```diff
@@ -94,11 +94,9 @@ static int address_new_static(Network *network, const char *filename, unsigned s
...
         address->section = TAKE_PTR(n);
 
-        r = ordered_hashmap_ensure_allocated(&network->addresses_by_section, &network_config_hash_ops);
-        if (r < 0)
-                return r;
-
-        r = ordered_hashmap_put(network->addresses_by_section, address->section, address);
+        r = ordered_hashmap_ensure_put(&network->addresses_by_section, &network_config_hash_ops, address->section, address);
+        if (r == -ENOMEM)
+                return log_oom();
```
Comment (yuwata) [inline src/network/networkd-address.c]: Please drop these two lines. `address_new_static()` should not log but simply return negative errno:
```
r = ordered_hashmap_ensure_put(...);
if (r < 0)
    return r;
```

File: src/network/networkd-network.c [hunk 4]
```diff
@@ -499,15 +499,13 @@ int network_load_one(Manager *manager, OrderedHashmap **networks, const char *fi
...
                 return 0;
 
-        r = ordered_hashmap_ensure_allocated(networks, &string_hash_ops);
-        if (r < 0)
-                return r;
-
-        r = ordered_hashmap_put(*networks, network->name, network);
+        r = ordered_hashmap_ensure_put(networks, &string_hash_ops, network->name, network);
+        if (r == -ENOMEM)
+                return log_oom();
```
Comment (yuwata) [inline src/network/networkd-network.c]: Same here. Please drop these two lines.

File: src/sysusers/sysusers.c [hunk 5]
```diff
@@ -1043,13 +1043,11 @@ static int add_user(Item *i) {
...
-        r = ordered_hashmap_ensure_allocated(&todo_uids, NULL);
-        if (r < 0)
+        r = ordered_hashmap_ensure_put(&todo_uids, NULL, UID_TO_PTR(i->uid), i);
+        if (r == -ENOMEM)
                 return log_oom();
-
-        r = ordered_hashmap_put(todo_uids, UID_TO_PTR(i->uid), i);
         if (r < 0)
-                return log_oom();
+                return r;
```
Comment (yuwata) [inline src/sysusers/sysusers.c]: Could you add error logs here?
```
if (r == -EEXIST)
    return log_error_errno(r, "Requested user %s (%s) with uid " UID_FMT " and gid "GID_FMT " to be created is duplicated or conflicts with another user.", ...);
if (r < 0)
    return log_error_errno(r, "Failed to store user %s (%s) with uid " UID_FMT " and gid " GID_FMT " to be created: %m");
```
This is not your fault in this PR, but you changed the code here. So, please also adjust it.

File: src/sysusers/sysusers.c [hunk 6]
```diff
@@ -1212,13 +1210,11 @@ static int add_group(Item *i) {
...
-        r = ordered_hashmap_ensure_allocated(&todo_gids, NULL);
-        if (r < 0)
+        r = ordered_hashmap_ensure_put(&todo_gids, NULL, GID_TO_PTR(i->gid), i);
+        if (r == -ENOMEM)
                 return log_oom();
-
-        r = ordered_hashmap_put(todo_gids, GID_TO_PTR(i->gid), i);
         if (r < 0)
-                return log_oom();
+                return r;
```
Comment (yuwata) [inline src/sysusers/sysusers.c]: Same here. Please also add error logs here.

File: src/udev/udev-rules.c [hunk 7]
```diff
@@ -1977,13 +1977,12 @@ static int udev_rule_apply_token_to_event(
...
-                r = ordered_hashmap_ensure_allocated(&event->seclabel_list, NULL);
-                if (r < 0)
+                r = ordered_hashmap_ensure_put(&event->seclabel_list, NULL, name, label);
+                if (r == -ENOMEM)
                         return log_oom();
-
-                r = ordered_hashmap_put(event->seclabel_list, name, label);
                 if (r < 0)
-                        return log_oom();
+                        return r;
```
Comment (yuwata) [inline src/udev/udev-rules.c]: Please also add  a error log here.
```
if (r < 0)
    return log_rule_error_errno(dev, rules, r, "Failed to store SECLABEL{%s}='%s': %m", name, label);
```

File: src/udev/udev-rules.c [hunk 8]
```diff
@@ -1977,13 +1977,12 @@ static int udev_rule_apply_token_to_event(
...
+
                 log_rule_debug(dev, rules, "SECLABEL{%s}='%s'", name, label);
                 name = label = NULL;
```
Comment (yuwata) [inline src/udev/udev-rules.c]: `TAKE_PTR()`?

**Review by yuwata — CHANGES_REQUESTED** [review]

**Comment by ssahani** [comment]
updated thanks for the review ! ptal

File: src/sysusers/sysusers.c [hunk 9]
```diff
@@ -1212,13 +1214,13 @@ static int add_group(Item *i) {
                 i->gid = search_uid;
         }
 
-        r = ordered_hashmap_ensure_allocated(&todo_gids, NULL);
-        if (r < 0)
+        r = ordered_hashmap_ensure_put(&todo_gids, NULL, GID_TO_PTR(i->gid), i);
+        if (r == -EEXIST)
+                return log_error_errno(r, "Requested gid "GID_FMT " to be created is duplicated or conflicts with another user.", i->gid);
```
Comment (yuwata) [inline src/sysusers/sysusers.c]: Please add group name in the log.
`Requested group %s with gid...`

File: src/sysusers/sysusers.c [hunk 10]
```diff
@@ -1212,13 +1214,13 @@ static int add_group(Item *i) {
...
+        if (r == -ENOMEM)
                 return log_oom();
-
-        r = ordered_hashmap_put(todo_gids, GID_TO_PTR(i->gid), i);
         if (r < 0)
-                return log_oom();
+                return r;
```
Comment (yuwata) [inline src/sysusers/sysusers.c]: `return log_error_errno(r, "Failed to store group %s with gid " GID_FMT " to be created: %m", i->name, i->gid);`

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/sysusers/sysusers.c (hunk 10 above)
Comment (ssahani) [reply src/sysusers/sysusers.c]: Updated thanks !

**Review by yuwata — APPROVED** [review]
Thanks. LGTM.

**Pull request merged**